### PR TITLE
Ruby final

### PR DESCRIPTION
**Pull request by Michelle-Lohwt**

### CODE REVIEW

File: sec2_the-ruby-programming-language/2.12_final_ruby_project/1_bcrypt.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
+require 'bcrypt'
+
+my_password = BCrypt::Password.create("my password")
+
+puts my_password
+puts my_password.version              #=> "2a"
```
Comment (Michelle-Lohwt) [inline sec2_the-ruby-programming-language/2.12_final_ruby_project/1_bcrypt.rb:6]: What is this `.version`?

File: sec2_the-ruby-programming-language/2.12_final_ruby_project/1_bcrypt.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+puts my_password.cost                 #=> 12
```
Comment (Michelle-Lohwt) [inline sec2_the-ruby-programming-language/2.12_final_ruby_project/1_bcrypt.rb:7]: What is `.cost`?

File: sec2_the-ruby-programming-language/2.12_final_ruby_project/2_bcrypt_autheticate.rb [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+# Hash the password
+def create_hash_digest(password)
+  BCrypt::Password.create(password)
+end
+
+# Verify the password
+def verify_hash_digest(password)
+  BCrypt::Password.new(password)
+end
```
Comment (Michelle-Lohwt) [inline sec2_the-ruby-programming-language/2.12_final_ruby_project/2_bcrypt_autheticate.rb:19]: What is the difference between `.create` and `.new` in Line 13, Line 18

File: sec2_the-ruby-programming-language/2.12_final_ruby_project/1_bcrypt.rb (hunk 1 above)
Comment (tlkiong) [reply sec2_the-ruby-programming-language/2.12_final_ruby_project/1_bcrypt.rb:6]: This is a method by the Bcrypt library. Go find their library and. comment here what it does.

File: sec2_the-ruby-programming-language/2.12_final_ruby_project/1_bcrypt.rb (hunk 2 above)
Comment (tlkiong) [reply sec2_the-ruby-programming-language/2.12_final_ruby_project/1_bcrypt.rb:7]: This is a method by the Bcrypt library. Go find their library and. comment here what it does.

File: sec2_the-ruby-programming-language/2.12_final_ruby_project/2_bcrypt_autheticate.rb (hunk 3 above)
Comment (tlkiong) [reply sec2_the-ruby-programming-language/2.12_final_ruby_project/2_bcrypt_autheticate.rb:19]: These are 2 methods in the bcrypt library.
`new` is `initialize` (unless they rewrite it). But create can be anything.
Please go to their docs / code and find out what each does and comment here :)

File: sec2_the-ruby-programming-language/2.12_final_ruby_project/3_create_module/main.rb [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+# $LOAD_PATH << "./<filename>"    // `.` is the universal sign for your current directory
+# require 'crud_module'
+
+users = [
+  {username: "michelle", password: "password1"},
+  {username: "apple", password: "password2"},
+  {username: "data", password: "password3"},
+  {username: "delta", password: "password4"},
+  {username: "alpha", password: "password5"},
+  ]
```
Comment (tlkiong) [inline sec2_the-ruby-programming-language/2.12_final_ruby_project/3_create_module/main.rb:14]: indentation ah...

File: sec2_the-ruby-programming-language/2.12_final_ruby_project/1_bcrypt.rb (hunk 1 above)
Comment (Michelle-Lohwt) [reply sec2_the-ruby-programming-language/2.12_final_ruby_project/1_bcrypt.rb:6]: `version` is an attribute in the module `BCrypt`, class `Password` where it shows the version of the bcrypt algorithm used to create a hash.
When `BCrypt::Password` is created (in Line 3), it initialise a hash and split the hash with `$` and get the version.

File: sec2_the-ruby-programming-language/2.12_final_ruby_project/1_bcrypt.rb (hunk 2 above)
Comment (Michelle-Lohwt) [reply sec2_the-ruby-programming-language/2.12_final_ruby_project/1_bcrypt.rb:7]: `.cost` is the cost needed to create a hash. The higher the cost, the harder attackers guess the password but also takes longer time for the computer to find the user's password.

Comment (tlkiong) [reply sec2_the-ruby-programming-language/2.12_final_ruby_project/1_bcrypt.rb:7]: Ok. So, how do you define cost in technical terms?

Comment (Michelle-Lohwt) [reply sec2_the-ruby-programming-language/2.12_final_ruby_project/1_bcrypt.rb:7]: cost factor

File: sec2_the-ruby-programming-language/2.12_final_ruby_project/2_bcrypt_autheticate.rb (hunk 3 above)
Comment (Michelle-Lohwt) [reply sec2_the-ruby-programming-language/2.12_final_ruby_project/2_bcrypt_autheticate.rb:19]: `.new`/ `.initialise` will check if the validity of the raw hash created, split it using `$` and return the separated hash while `.create` will calculate the cost factor of the hash and create a hash and pass it to `.initialise`.

**Comment by tlkiong** [comment]
image

you say me previously for commenting it when its still in draft. Yet now you ask my review when its still in draft. So, is it ready for review?
🤔

**Comment by Michelle-Lohwt** [comment]
Sorry, not yet hahahahh

**Review by tlkiong — CHANGES_REQUESTED** [review]
Indentation.....

**Review by tlkiong — CHANGES_REQUESTED** [review]
Please double triple check your work before asking me to review.

**Review by tlkiong — APPROVED** [review]

**Pull request merged**